### PR TITLE
Multigrid merge: part 4

### PR DESCRIPTION
**Pull request by ThomasCeulemans**
Final part of the multigrid merge:

Includes all methods to use the multi-resolution methods.
Also included multi-resolution examples of 'standard' test scripts.
Cleaned up the voro++ dependency a bit with the fork.
Now also removed all references to multigrid to avoid confusion (because the thing we implement is not exactly multigrid).

### CODE REVIEW

File: src/bindings/pybindings.cpp [hunk 1]
```diff
@@ -118,9 +118,13 @@ PYBIND11_MODULE (core, module)
         .def ("compute_Jeff",                                                       &Model::compute_Jeff)
         .def ("compute_level_populations_from_stateq",                              &Model::compute_level_populations_from_stateq)
         .def ("compute_level_populations",                                          &Model::compute_level_populations)
+        .def ("compute_level_populations_multiresolution",                                &Model::compute_level_populations_multiresolution)
```
Comment (FredDeCeuster) [inline src/bindings/pybindings.cpp]: Wrong indentation format.
Please align in columns.

File: src/bindings/pybindings.cpp [hunk 2]
```diff
@@ -118,9 +118,13 @@ PYBIND11_MODULE (core, module)
...
         .def ("compute_image",                                                      &Model::compute_image)
         .def ("set_eta_and_chi",                                                    &Model::set_eta_and_chi)
         .def ("set_boundary_condition",                                             &Model::set_boundary_condition)
+        .def ("setup_multiresolution",                                                    &Model::setup_multiresolution)
```
Comment (FredDeCeuster) [inline src/bindings/pybindings.cpp]: Please align in columns.

File: src/model/mrController/vCycle/vCycle.hpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    Size finest_lvl=0;//Debug variable, denotes the finest level the multiresolution sequence should go to TODO DEPRECATE THIS
+    Size max_level=0;//number of levels going from 0 (finest grid) to n_levels-1 (coarsest grid)
+    Size current_level=0;//the current level
+    Size n_pre_interpolation_steps=1;//number of iterations on the coarsest grid before interpolating
+
+    Size max_n_iterations;//the maximum number of cycles allowed for this multiresolution scheme
+    Size current_n_iterations=0;
+
+    bool is_next_action_set=false;//checks whether the next action has been set
+    Actions next_action;//the next action (if it has been set)
```
Comment (FredDeCeuster) [inline src/model/mrController/vCycle/vCycle.hpp]: Please try column alignment.
```
type1   var1   = val1;     // Comment1
type11  var11  = val11;    // Comment11
type111 var111 = val111;   // Comment111
```

File: src/model/mrController/vCycle/vCycle.tpp [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    {
+        return Actions::finish;
+    }
+
+    //If converged, then use whatever recommended by converged_on_current_grid
+    if (is_next_action_set)
+    {
+        is_next_action_set=false;
+        return next_action;
+    }else{
```
Comment (FredDeCeuster) [inline src/model/mrController/vCycle/vCycle.tpp]: Be consistent in formatting.
```
if ()
{
}
else
{
}
```

File: src/model/mrController/wCycle/wCycle.hpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    Size max_n_iterations;//the maximum number of cycles allowed for this multiresolution scheme
+    Size current_n_iterations=0;
+
+    bool is_next_action_set=false;//checks whether the next action has been set
+    Actions next_action;//the next action (if it has been set)
+    //for now, I implement the V-cycle
+
+    vector<Actions> action_order;//contains the order of the actions (interpolate_corrections and restrict) in the w-cycle
+    //the state for the multiresolution cycle was already getting complicated for the v-cycle, ergo this helps alleviate it
+    Size curr_action_nb=0;// the index at which we currently are in the action_order
```
Comment (FredDeCeuster) [inline src/model/mrController/wCycle/wCycle.hpp]: Columize.

File: tests/CMakeLists.txt [hunk 6]
```diff
@@ -104,7 +106,7 @@ target_link_libraries (test_solver_lambda Magritte)
 
 if (OpenMP_CXX_FOUND)
     target_link_libraries (test_raytracer         OpenMP::OpenMP_CXX)
-    target_link_libraries (test_multigrid         OpenMP::OpenMP_CXX)
+    target_link_libraries (test_multiresolution         OpenMP::OpenMP_CXX)
```
Comment (FredDeCeuster) [inline tests/CMakeLists.txt]: columize

File: tests/CMakeLists.txt [hunk 7]
```diff
@@ -115,15 +117,15 @@ endif()
 if (OMP_PARALLEL)
     if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
         target_link_libraries (test_raytracer         atomic)
-        target_link_libraries (test_multigrid         atomic)
+        target_link_libraries (test_multiresolution         atomic)
```
Comment (FredDeCeuster) [inline tests/CMakeLists.txt]: columize

File: tests/CMakeLists.txt [hunk 8]
```diff
@@ -115,15 +117,15 @@ endif()
...
         target_link_libraries (test_levelpops         atomic)
         target_link_libraries (test_shortchar_order_0 atomic)
         target_link_libraries (test_feautrier_order_2 atomic)
         target_link_libraries (test_solver_lambda     atomic)
         target_link_libraries (test_imager            atomic)
     else ()
         target_link_libraries (test_raytracer         OpenMP::OpenMP_CXX)
-        target_link_libraries (test_multigrid         OpenMP::OpenMP_CXX)
+        target_link_libraries (test_multiresolution         OpenMP::OpenMP_CXX)
```
Comment (FredDeCeuster) [inline tests/CMakeLists.txt]: columize

File: tests/benchmarks/numeric/vanZadelhoff_1_3D_mesher.py [hunk 9]
```diff
@@ -54,8 +54,8 @@
 mesh = mesher.Mesh(meshName)
 
 npoints = len(mesh.points)
-# nbs     = [n for sublist in mesh.neighbors for n in sublist]
-# n_nbs   = [len(sublist) for sublist in mesh.neighbors]
+nbs     = [n for sublist in mesh.neighbors for n in sublist]
+n_nbs   = [len(sublist) for sublist in mesh.neighbors]
```
Comment (FredDeCeuster) [inline tests/benchmarks/numeric/vanZadelhoff_1_3D_mesher.py]: Do not use the neighbors derived form the `magritte.mesher.Mesh`.
There are models for which this does not work.
Instead use or adapt the `magritte.setup.set_Delaunay_neighbor_lists` function.
This basically re-Delaunay-tetrahedralizes the given point cloud, but has a better behaved neighbour structure than our derived one.

File: tests/benchmarks/numeric/vanZadelhoff_1_3D_mesher.py [hunk 10]
```diff
@@ -91,6 +91,7 @@ def nTT (r):
     model.geometry.points.position.set(mesh.points)
     model.geometry.points.velocity.set(np.zeros((npoints, 3)))
 
+    model.geometry.points.multiscale.set_all_neighbors(n_nbs,nbs)
```
Comment (FredDeCeuster) [inline tests/benchmarks/numeric/vanZadelhoff_1_3D_mesher.py]: Perhaps, move to `magritte.setup.set_Delaunay_neighbor_lists` cfr. previous comment.

File: tests/benchmarks/numeric/vanZadelhoff_1_3D_mesher.py (hunk 9 above)
Comment (FredDeCeuster) [reply tests/benchmarks/numeric/vanZadelhoff_1_3D_mesher.py]: Take this into account in all test scripts.

File: tests/benchmarks/numeric/vanZadelhoff_1_3D_mesher_multiresolution.py [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+                         radius_in  = r_in,
+                         radius_out = r_out),
+    scale_min      = scale_min,
+    scale_max      = scale_max,
+    scale_function = scale_fun )
+
+mesh = mesher.Mesh(meshName)
+
+npoints = len(mesh.points)
+nbs     = [n for sublist in mesh.neighbors for n in sublist]
```
Comment (FredDeCeuster) [inline tests/benchmarks/numeric/vanZadelhoff_1_3D_mesher_multiresolution.py]: Same as before.

**Review by FredDeCeuster — CHANGES_REQUESTED** [review]
Please address the following comments.

File: tests/benchmarks/numeric/vanZadelhoff_1_3D_mesher.py (hunk 9 above)
Comment (ThomasCeulemans) [reply tests/benchmarks/numeric/vanZadelhoff_1_3D_mesher.py]: This is now changed for all test scripts.

File: src/bindings/pybindings.cpp (hunk 1 above)
Comment (ThomasCeulemans) [reply src/bindings/pybindings.cpp]: Most indentation in the pybindings and .hpp files should now be aligned properly.

**Pull request merged**